src/handlers.cpp

```diff
@@ -1579,28 +1632,129 @@ void ImpuRegDataTask::on_sar_response(Diameter::Message& rsp)
         TRC_DEBUG("Getting User-Data from SAA for cache");
         saa.user_data(_xml);
         put_in_cache();
+        caching = true;
       }
-      send_reply();
       break;
     case 5001:
     {
       TRC_INFO("Server-Assignment answer with result code %d - reject", result_code);
       SAS::Event event(this->trail(), SASEvent::REG_DATA_HSS_FAIL, 0);
       SAS::report_event(event);
-      send_http_reply(HTTP_NOT_FOUND);
+      _http_rc = HTTP_NOT_FOUND;
     }
     break;
     default:
       TRC_INFO("Server-Assignment answer with result code %d - reject", result_code);
       SAS::Event event(this->trail(), SASEvent::REG_DATA_HSS_FAIL, 0);
       SAS::report_event(event);
-      send_http_reply(HTTP_SERVER_ERROR);
+      _http_rc = HTTP_SERVER_ERROR;
       break;
   }
-  delete this;
+
+  // If we're not caching the registration data, check for deregistration and
+  // delete the task now
+  if (!caching)
```

Comment:
Ah - I remember.  We only call put_in_cache at line 1634 if is_deregistration_request(_type) is false, so we don't need to do the code from 1662 to 1694 if we've done a put_in_cache.  Does that make sense?  Do you want me to change the commenting to make this clearer?